### PR TITLE
Feature/cdap 1367 namespaced streams metrics

### PR DESCRIPTION
**Pull request by anwar6953**
Namespaced stream metrics

Build: http://builds.cask.co/browse/CDAP-DUT825-8

### CODE REVIEW

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTest.java [hunk 1]
```diff
@@ -58,9 +58,9 @@
     new GsonBuilder().registerTypeAdapter(Schema.class, new SchemaTypeAdapter())).create();
 
 
-  protected abstract URL constructPath(String path) throws URISyntaxException, MalformedURLException;
+  protected abstract URL createURL(String path) throws URISyntaxException, MalformedURLException;
 
-  private HttpURLConnection openURL(URL url, HttpMethod method) throws IOException {
+  protected HttpURLConnection openURL(URL url, HttpMethod method) throws IOException {
```
Comment (gsps1) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTest.java]: why protected ? i guess its used only in this file ?

Comment (anwar6953) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTest.java]: It's also used in StreamHandlerTestV2 and StreamHandlerTestV3

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java [hunk 2]
```diff
@@ -17,17 +17,163 @@
...
+        return;
+      }
+      TimeUnit.SECONDS.sleep(1);
+    }
+    Assert.assertEquals(expectedCount, getNumProcessed(streamId));
+  }
+
+
+  private int getNumProcessed(Id.Stream streamId) throws Exception {
+    String path = String.format("/v3/metrics/query?metric=system.collect.events&context=ns.%s.str.%s",
```
Comment (gsps1) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java]: pull latest develop, after https://github.com/caskdata/cdap/pull/1365 
we use user-friendly tag name now, so it would be like , namespace.%s.stream.%s

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java [hunk 3]
```diff
@@ -17,17 +17,163 @@
...
+                                streamId.getNamespaceId(), streamId.getName());
+    System.out.println(path);
```
Comment (gsps1) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java]: remove sys-outs.

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java [hunk 4]
```diff
@@ -17,17 +17,163 @@
...
+    HttpRequest request = HttpRequest.post(getEndPoint(path).toURL()).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+    System.out.println(response.getResponseBodyAsString());
```
Comment (gsps1) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java]: same

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java [hunk 5]
```diff
@@ -17,17 +17,163 @@
...
+    return getNumEventsFromResponse(response.getResponseBodyAsString());
+  }
+
+  private int getNumEventsFromResponse(String response) {
+    // response format:
+    // [{"measureName":"system.collect.events","tagValues":{},"timeValues":[{"timestamp":0,"value":12}]}]
+
+    JsonElement jsonResponse = new JsonParser().parse(response);
```
Comment (gsps1) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java]: i would suggest using gson here,
```new Gson().fromJson(EntityUtils.toString(response.getEntity()), MetricQueryResult.class);```
and then read the MetriQueryResult for the value.

**Comment by gsps1** [comment]
@anwar6953 reviewed the metrics part, couple of minor comments, looks good otherwise.

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java (hunk 5 above)
Comment (anwar6953) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java]: thanks for the advice

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java (hunk 2 above)
Comment (anwar6953) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java]: thanks, it worked.

**Comment by anwar6953** [comment]
@gsps1 thanks; updated.

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java [hunk 6]
```diff
@@ -298,7 +298,7 @@
      */
     public static final String API_VERSION_3_TOKEN = "v3";
     public static final String API_VERSION_3 = "/" + API_VERSION_3_TOKEN;
-    public static final String STREAM_HANDLER_NAME = "stream.rest";
+    public static final String STREAM_HANDLER_NAME = "stream_rest";
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: why this change?

**Comment by bdmogal** [comment]
1 Minor comment, but LGTM otherwise

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java (hunk 6 above)
Comment (anwar6953) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: due to how queries happen (using period in the queryParams), the metrics Query handler splits on period, so this messes it up.

**Comment by gsps1** [comment]
LGTM

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java (hunk 6 above)
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: ah ok!
